File: POO_2026_1/src/POO/Animal.java
package POO;

import MeuPacote.Geral;
import java.util.ArrayList;

public abstract class Animal {
  private String nome;

    public Animal(String nome) {
        this.nome = nome;
    }

    public String getNome() {
        return nome;
    }

    public void setNome(String nome) {
        this.nome = nome;
    }
    
    public abstract void som();
    
    public static void listarAnimais(ArrayList<Animal> meusAnimais) {
      for(Animal item: meusAnimais) {
        System.out.printf("%s\n", item.toString());         
        item.som();
        if (item instanceof Cachorro)
            ((Cachorro)item).rosnar();
        System.out.println(Geral.repetir("=", 30));
      }  
      
      System.out.printf("Temos %d Cachorro(s)\n", Cachorro.ct);
      System.out.printf("Temos %d Gato(s)\n", Gato.ct);
      System.out.printf("Temos %d Vaca(s)\n", Vaca.ct);
    }

}

File: CPP/POO_CPP/MundoAnimal/Animal.cpp
#include <iostream>
#include "Animal.h"
#include "Cachorro.h"
#include "Gato.h"
#include "Vaca.h"

using namespace std;

Animal::Animal(string nome) {
  this->nome = nome;
}

Animal::~Animal() {}

string Animal::getNome() {
  return nome;
}

void Animal::setNome(string nome) {
  this->nome = nome;
}

void Animal::mostrarAnimais(vector<Animal *> x) {
  for(Animal *item: x) {
    cout << item->toString() << endl;
    item->som();

    Cachorro *ptCachorro = dynamic_cast<Cachorro *>(item);
    if (ptCachorro) {
        ptCachorro->rosnar();
    }

    cout << "==============================================" << endl;
  }
  cout << "Temos " << Cachorro::ct << " Cachorro(s)." << endl;
  cout << "Temos " << Gato::ct << " Gato(s)." << endl;
  cout << "Temos " << Vaca::ct << " Vaca(s)." << endl;
}

void Animal::som() {}
string Animal::toString() {}
